### ISSUE TITLE
Strava heatmap authentication

### ISSUE THREAD

**Issue by bertt** [bertt/wmts#2]
Strava heatmaps after level 12 give now an error, for example:

```
$ wget https://heatmap-external-c.strava.com/tiles/all/hot/14/8423/5363.png
authentication is now required for data access beyond zoom level 12, please login to strava.com and then visit /auth
```

How to add this authentication stuff? Did a quick test retrieving tile with added tokens but no luck yet.

**Comment by alpter** [comment]
When used under WMTS on QGIS with 
https://bertt.github.io/wmts/capabilities/strava.xml
I can see it between 0 and 13 zoom... From 14 onwards just disappear.... it does not state any errors ... just dont show it...

Used to work a couple of days ago thou....

**Comment by bertt** [comment]
yes here the blog describing the changes: https://blog.strava.com/press/heatmap-updates/

**Comment by bertt** [comment]
it seems also using the website it's not possible to get to the detailed heatmap tiles.

- Go to https://www.strava.com/heatmap#7.00/-119.32346/37.73905/hot/all and login/zoomin
- Tiles are retrieved from: https://heatmap-external-c.strava.com/tiles-auth/all/hot/6/12/23.png?v=19
response: 
```
<Error>
<Code>AccessDenied</Code>
<Message>Access denied</Message>
</Error>
```

**Comment by bertt** [comment]
on site https://www.strava.com/heatmap#18.00/-119.26759/36.19952/blue/all I see a lot of 403 errors (Access denied) on detailed tile levels and after login.

Tile are retrieved from tiles-auth route: https://heatmap-external-b.strava.com/tiles-auth/all/blue/16/11054/25688.png?v=19

**Comment by alpter** [comment]
@mbernasocchi  any idea ?  issue above ?
This seems still to be working ....
http://globalheat.strava.com/tiles/both/color1/{z}/{x}/{y}.png

> Title:	Strava Both
> Credit:	© Strava for tracing into OSM only
> URL:	http://globalheat.strava.com/tiles/both/color1/{z}/{x}/{y}.png
> yOrigin:	Top (yOriginTop=1)
> Zoom range:	0 - 18
> Layer Extent:	Not set

It is coming form NextGIS...  @simgislab  any idea the issue above ?
https://github.com/nextgis/quickmapservices_contrib

**Comment by alpter** [comment]
https://github.com/strava/ec2metaproxy/pull/1

**Comment by simgislab** [comment]
@alpter don't use contrib, rarely updated, use http://qms.nextgis.com, search for Strava, may be other sources would help better

**Comment by bertt** [comment]
the old heatmap (http://globalheat.strava.com/tiles/both/color1/0/0/0.png) still works but has less features and is not updated anymore i guess.

**Comment by andrea-del-sarto** [comment]
hi, you can try this https://heatmap-external-b.strava.com/tiles-auth/all/bluered/{zoom}/{x}/{y}.png?px=256
It should work

**Comment by bertt** [comment]
@andrea-del-sarto it gives an error, for example: 
```
$ curl https://heatmap-external-b.strava.com/tiles-auth/all/bluered/14/8423/5363.png?px=256
<?xml version="1.0" encoding="UTF-8"?><Error><Code>MissingKey</Code><Message>Missing Key-Pair-Id query parameter or cookie value</Message></Error>
```

Question is: How to add a 'Key-Pair-Id query parameter'?

**Comment by alpter** [comment]
@andrea-del-sarto does not work....
I found this https://github.com/osmlab/editor-layer-index/pull/373
on the last exchange they talk about  "maximum zoom level" permitted and "authentication"...

**Comment by alpter** [comment]
@tyrasd  it seems we can not get the old version of the strava heatmap anymore below zoom 12/13.
Is there any solution ?  In JOSM same problem...
https://bertt.github.io/wmts/capabilities/strava.xml

**Comment by alpter** [comment]
I guess we are stacked at present... until STRAVA  / MAPBOX wont solve it...
https://josm.openstreetmap.de/ticket/16100

**Comment by andrea-del-sarto** [comment]
sorry, i've use it on osm id/strava id how custom background with strava login :(
anyway zoom level 12 work if use 256px version of the tiles
don't work > https://heatmap-external-c.strava.com/tiles/both/bluered/12/1505/2306.png
work > https://heatmap-external-c.strava.com/tiles/both/bluered/12/1505/2306.png?px=256

**Comment by alpter** [comment]
... see my comment here
> > exactly!Since the new Heat Map is being developed under[​Mapbox GL](https://www.mapbox.com/help/define-mapbox-gl/) they just need to set proper individual ​[access tokens](https://www.mapbox.com/help/define-access-token/) in order to associate requests to API resources with each personal STRAVA account. I guess... but... How long would it take ?
> https://josm.openstreetmap.de/ticket/16100

**Comment by mourner** [comment]
@alpter as far as I know, Strava has their own heatmap implementation that is completely independent of ours. No idea how to help you here.

**Comment by tyrasd** [comment]
me neither, but maybe (hopefully) someone from @strava can tell us more about this issue, e.g. @drewrobb.

**Comment by scrapperuk** [comment]
keen to find a fix for this as an end user... I have a Strava login but cant apply the authentication to the iD editor..
I've even tried contacting the guy who developed this while at Strava, as its too cool to lose

will let you know if i hear anything (although I'm not a coder, so might have to use some plain English)

**Comment by alpter** [comment]
@inexcitus could you have a look at this issue... any idea ?

**Comment by alpter** [comment]
Thanks @scrapperuk  lets keep it up..
https://github.com/paulmach/slide/issues/12

**Comment by StevenClays** [comment]
@bertt In JOSM its works, adding the cookies values by doing something like this > 
tms[3,15]:https://heatmap-external-{switch:a,b,c}.strava.com/tiles-auth/both/bluered/{zoom}/{x}/{y}.png?Key-Pair-Id={my-value}&Signature={my-value}&Policy={my-value}
But how can I achieve the same result in QGIS? Thanks for helping me out

**Comment by bertt** [comment]
@StevenClays  ah interesting, so only with cookie parameters in the request querystring (and not in the request headers)?

**Comment by StevenClays** [comment]
In JOSM this works, but I cannot get QGIS to do this. What could I be doing wrong? I'm not so acquainted with terms as 'request headers', in JOSM I just entered this link and it worked...

**Comment by bertt** [comment]
@StevenClays  where did you get that url from?

**Comment by StevenClays** [comment]
@bertt In the JOSM-thread regarding this issue > https://josm.openstreetmap.de/ticket/16100

**Comment by bertt** [comment]
ok thanks :-) At the moment I don't know why QGIS doesn't work with this url but JOSM apparently does work.

**Comment by StevenClays** [comment]
@bertt Thanks for looking into it. Hopefully you can work it out...

**Comment by luiswoo** [comment]
Hi, is it possible to add gps tracks by Yandex (n.maps.yandex.ru)?
https://core-gpstiles.maps.yandex.net/tiles?style=point&x=37275&y=19294&z=16
https://core-gpstiles.maps.yandex.net/tiles?style=red_combined&x=37268&y=19297&z=16
```
"ygpx": { \
     "name": "yandex gps point",
     "prefix": "ygpx",			# tile directory
     "ext": "png",			# tile images extension
     "scalable": False,			# could zN tile be constructed of four z(N+1) tiles
     "fetch": fetchers.Tile,		# function that fetches given tile. should return None if tile wasn't fetched
     "remote_url": "https://gps-tiles.maps.yandex.net/tiles?style=point&x=%s&y=%s&z=%s",
     "transform_tile_number": lambda z,x,y: (x,y,z-1),
     "dead_tile": install_path + "yandex_nxt.jpg",
     "proj": "EPSG:3395",
},\
"ygpx_red_combined": { \
     "name": "yandex gps red_combined",
     "prefix": "ygpx_red_combined",		# tile directory
     "ext": "png",				# tile images extension
     "scalable": False,				# could zN tile be constructed of four z(N+1) tiles
     "fetch": fetchers.Tile,			# function that fetches given tile. should return None if tile wasn't fetched
     "remote_url": "https://gps-tiles.maps.yandex.net/tiles?style=red_combined&x=%s&y=%s&z=%s",
     "transform_tile_number": lambda z,x,y: (x,y,z-1),
     "dead_tile": install_path + "yandex_nxt.jpg",
     "proj": "EPSG:3395",
},\
```
 ^
this is for TWMS https://github.com/Komzpa/twms

https://n.maps.yandex.ru/#!/help?z=16&ll=24.740202%2C59.432151&l=nk%23sat

**Comment by nnngrach** [comment]
I have written a script to automatically log in to Strava and receiving of cookies. To do this, I used a Headless Chrome browser controlled by the Puppeteer script. Details and the script on this link:

**Comment by Jochem-L** [comment]
> @bertt In JOSM its works, adding the cookies values by doing something like this >
> tms[3,15]:[https://heatmap-external-{switch:a,b,c}.strava.com/tiles-auth/both/bluered/{zoom}/{x}/{y}.png?Key-Pair-Id={my-value}&Signature={my-value}&Policy={my-value}](https://heatmap-external-%7Bswitch:a,b,c%7D.strava.com/tiles-auth/both/bluered/%7Bzoom%7D/%7Bx%7D/%7By%7D.png?Key-Pair-Id=%7Bmy-value%7D&Signature=%7Bmy-value%7D&Policy=%7Bmy-value%7D)
> But how can I achieve the same result in QGIS? Thanks for helping me out

Did anyone got it working in QGIS?

Simply adding the URL in XYZ Tiles (after copying in Key-Pair-Id, Policy and Signature values) seems not to be working:
https://heatmap-external-b.strava.com/tiles-auth/all/bluered/{z}/{x}/{y}.png?v=19&Key-Pair-Id=APKAID...&Policy=eyJTdG...&Signature=mpbSxZ...

**Comment by nnngrach** [comment]
Hi. I updated the API for my service to automatically access Strava tiles without authorization and replacing cookies. You can just use a URL like this: 

https://anygis.ru/api/v1/Tracks_Strava_All/{x}/{y}/{z}

The full list of available URL maps can be found here:

https://github.com/nnngrach/strava_auto_auth

**Comment by KommandorKeen** [comment]
Fantastic, Jochem's solution worked for me. I have just copied the instructions into the layers metadata in QGIS in case I forget!
![image](https://user-images.githubusercontent.com/16811869/86881666-29152a80-c132-11ea-89e2-0f0e577b8224.png)

**Comment by forrester86** [comment]
Help, 
strava cookies no longer contain these parameters!
CloudFront-Key-Pair-Id
CloudFront-Policy
CloudFront-Signature
How do I get access now?

**Comment by pdxvr** [comment]
I just cleared my Strava cookies and reloaded the page and I have all of those cookies back.  I’m not sure what’s going on for you, but I still have them.

**Comment by forrester86** [comment]
Everything is working!   The keys are apparently updated not immediately, and not after an hour, but after a few hours.

**Comment by Markus93-code** [comment]
Help, how can I import https://anygis.ru/api/v1/Tracks_Strava_All/{x}/{y}/{z} into QGis? Thanks

**Comment by dkselw** [comment]
> 
> 
> > @Jochem-L I got it working in QGIS:
> 
> This has been working for me until recently. Now QGIS is logging the following error: WARNING Network request https://heatmap-external-a.strava.com/tiles-auth/both/gray/8/188/141.png?v=19&Key-Pair-Id= ... timed out
> 
> I can get it working with JOSM, but not QGIS. I also tried getting the cookie info with the Firefox and Chrome plugins from here: https://github.com/zekefarwell/josm-strava-heatmap

And now 4 days later I opened QGIS and the heatmap appeared without changing anything.

**Comment by dkselw** [comment]
The manual method of creating the URL described by @Jochem-L is working for me again.

The Firefox and Chrome plugins from here: https://github.com/zekefarwell/josm-strava-heatmap are also working for me to get the URL with cookie info:
I changed the start of the line from:
tms:https://heatmap-external-{switch:a,b,c}.strava.com/tiles-auth/all/hot/{zoom}/{x}/{y}.png
to
https://heatmap-external-a.strava.com/tiles-auth/all/hot/{z}/{x}/{y}.png

**Comment by Jochem-L** [comment]
> Hi. I updated the API for my service to automatically access Strava tiles without authorization and replacing cookies. You can just use a URL like this:
> 
> [https://anygis.ru/api/v1/Tracks_Strava_All/{x}/{y}/{z}](https://anygis.ru/api/v1/Tracks_Strava_All/%7Bx%7D/%7By%7D/%7Bz%7D)

@nnngrach, It seems like your service has been down for quite some time. It would be great if you could fix that. In any case, thanks for keeping this service up for so long. It gave me a lot of valuable information during hikes and I am missing this dearly.

**Comment by patrickziegler** [comment]
Hi guys, I had the same problem and found lots of helpful pointers in this thread, thanks a lot :+1: you might be interested in checking out https://github.com/patrickziegler/strava-heatmap-proxy

**Comment by Jochem-L** [comment]
Alternatively, to get Strava Heatmap in QGIS without installing a proxy server
 
- In Chrome, go to https://www.strava.com/heatmap
 
- Log in;
- Right-click the Strava logo and choose "Inspect";
- In top menu of DevTools, choose "Application";
- In left panel, fold out "Cookies" and click strava.com;
- Copy values of "CloudFront-Key-Pair-Id", "CloudFront-Policy" and "CloudFront-Signature";
- Paste these values to a text editor;
- In the text editor, copy-paste this URL:
 `https://heatmap-external-a.strava.com/tiles-auth/all/hot/{z}/{x}/{y}.png?v=19&Key-Pair-Id=...&Policy=...&Signature=...`
- Replace the `...` of "Key-Pair-Id", "Policy" and "Signature values" in the URL with your values;
- Optionally, replace the `all` in the URL with `ride`, `run`, `water` or `winter`;
- Optionally, replace the `hot` in the URL with `blue`, `purple`, `gray` or `bluered`;
- In QGIS, in browser panel, right click, "Choose New connection";
- In dialog window, type a free to choose name, paste URL and click "OK";
- In browser panel, double click the name;
- Repeat procedure when cookie is expired.

**Comment by patrickziegler** [comment]
@Jochem-L you are right, the sequence you posted above gave me the idea to add a slightly redesigned tool described in https://github.com/patrickziegler/strava-heatmap-proxy#using-the-static-file-configurator

Your example above would then boil down to the following line, which will replace all `%CloudFront-*%` tokens right away and print out the line you need for the connection in QGIS, no proxy needed

```sh
echo "https://heatmap-external-a.strava.com/tiles-auth/all/hot/%7Bz%7D/%7Bx%7D/%7By%7D.png?v=19&Key-Pair-Id=%CloudFront-Key-Pair-Id%&Policy=%CloudFront-Policy%&Signature=%CloudFront-Signature%" | strava-heatmap-auth
```

**Comment by patrickziegler** [comment]
@Jochem-L thanks for trying it

that config file is described in https://github.com/patrickziegler/strava-heatmap-proxy#wrench-configuration it holds the strava login credentials in the form
```sh
{"Email": "...", "Password": "..."}
```
with that in place, you can use it in a custom location with
```sh
echo "..." | strava-heatmap-auth --config "<path-to-your-config>"
```

**Comment by Jochem-L** [comment]
@patrickziegler, Your tool works for me only if I login to Strava manually first. It seems to use old Cloud-front values, if I don't. I expected the tool to do the new login too, since it needs the config file with login credentials. Do you have any ideas what goes wrong? Note that I'm running it under Windows 10 at the moment, but I might try Android 13 later too.

**Comment by patrickziegler** [comment]
@Jochem-L the configurator `strava-heatmap-auth` will perform a login on every startup (using the credentials you provide in the config). This is completely independent from what you do in your browser, the tool does not even look at any of your systems files or caches (apart from the config file of course).

However, as stated in the readme, the cloudfront tokens are expiring after around a week and you will need to manually re-create your TMS files accordingly after that time period.

Feel free to open an issue here https://github.com/patrickziegler/strava-heatmap-proxy/issues if you discover a bug, it would be highly appreciated, but I guess @bertt will not be happy if we continue discussing these things in his thread :smile:
